File: backend/app/services/benchmarks.py
"""The benchmark suite runner (AGENTS.md §13, KPIs in §13.1).

A suite is a YAML file of cases; a case is a prompt plus the expectations its run has to
meet. Running one means running the *whole graph* — plan, research, code, execute, debug,
track, evaluate, report — and then scoring the final state against those expectations. It
is the only test in this repository that exercises the real system end to end, which is why
§12 schedules it nightly rather than on PRs.

**Scoring is arithmetic over the final state, and every check is recorded individually.**
A case that reached `SUCCEEDED` but missed its accuracy floor and a case that crashed both
score `passed = false`; a scorecard that cannot tell them apart cannot tell a regression
from an outage. `checks` keeps them distinguishable, in the row and in the Markdown.

**The three trap cases are scored separately** (§13). `imbalance-trap`, `leakage-trap` and
`impossible-target` test judgement rather than capability: does the system notice a
misleading metric, catch a leaking feature, and fail honestly instead of reporting a
fabricated success? A platform that scores 7/7 on the ordinary cases and 0/3 on these is
not trustworthy, so the Judgement Score is its own line in the scorecard and never averaged
into the headline number.

**`report_mentions` is satisfied by any one entry, not all of them.** The strings are
alternative phrasings of the same idea — "not achievable", "limitation" — and a generative
model that says "cannot be attained with this data" has demonstrated exactly the honesty
the case is testing. Requiring every phrase would measure phrasing luck rather than
judgement, and a benchmark that fails for the wrong reason gets ignored.
"""

from __future__ import annotations

import asyncio
import logging
import math
import uuid
from dataclasses import dataclass, field
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

import yaml

from app.core.config import REPO_ROOT
from app.db.models.benchmark_result import BenchmarkResult
from app.db.models.task import Task, TaskStatus
from app.engine.criteria import COMPARATORS
from app.engine.state import AgentState, RunOutcome
from app.schemas.metrics import observed_metrics

logger = logging.getLogger(__name__)

SUITE_SUFFIX = ".yaml"

# Deliverable types a case may require in `expect.artifacts`.
KNOWN_ARTIFACT_TYPES = frozenset(
    {"code", "model", "plot", "report", "metrics", "log", "bundle"}
)


class SuiteNotFound(LookupError):
    """No suite file by that name."""


def benchmarks_root(root: Path | None = None) -> Path:
    """The `benchmarks/` directory at the repository root.

    Derived from `REPO_ROOT` rather than declared as a setting: unlike `/datasets` and
    `/runs`, which are Docker volumes whose mount point genuinely varies per environment,
    the suites are source files that live beside the code that reads them and are versioned
    with it. A configurable path would only ever be pointed somewhere else by mistake.
    """
    return root or (REPO_ROOT / "benchmarks")


def suites_root(root: Path | None = None) -> Path:
    return benchmarks_root(root) / "suites"


def results_root(root: Path | None = None) -> Path:
    return benchmarks_root(root) / "results"


# ------------------------------------------------------------------------------------
#  Suite definitions
# ------------------------------------------------------------------------------------


@dataclass(frozen=True)
class BenchmarkCase:
    id: str
    prompt: str
    task_kind: str = ""
    expect: dict[str, Any] = field(default_factory=dict)
    tests: str = ""
    trap: bool = False


@dataclass(frozen=True)
class BenchmarkSuite:
    name: str
    description: str
    version: str
    cases: list[BenchmarkCase]

    def case(self, case_id: str) -> BenchmarkCase | None:
        return next((c for c in self.cases if c.id == case_id), None)


def load_suite(name: str, root: Path | None = None) -> BenchmarkSuite:
    """Parse `benchmarks/suites/{name}.yaml`."""
    # A suite name reaches this function from a URL path parameter, so it is constrained to
    # a bare file stem before it is joined to a directory.
    if not name or "/" in name or "\\" in name or name.startswith("."):
        raise SuiteNotFound(f"'{name}' is not a valid suite name.")

    path = suites_root(root) / f"{name}{SUITE_SUFFIX}"
    try:
        payload = yaml.safe_load(path.read_text(encoding="utf-8"))
    except FileNotFoundError as exc:
        raise SuiteNotFound(f"No benchmark suite '{name}' at {path}.") from exc

    if not isinstance(payload, dict) or not isinstance(payload.get("cases"), list):
        raise SuiteNotFound(f"Benchmark suite '{name}' is not in the expected shape.")

    cases = [
        BenchmarkCase(
            id=str(entry["id"]),
            prompt=str(entry.get("prompt") or "").strip(),
            task_kind=str(entry.get("task_kind") or ""),
            expect=dict(entry.get("expect") or {}),
            tests=str(entry.get("tests") or ""),
            trap=bool(entry.get("trap", False)),
        )
        for entry in payload["cases"]
        if isinstance(entry, dict) and entry.get("id")
    ]
    return BenchmarkSuite(
        name=str(payload.get("suite") or name),
        description=str(payload.get("description") or "").strip(),
        version=str(payload.get("version") or "0.0.0"),
        cases=cases,
    )


def list_suites(root: Path | None = None) -> list[BenchmarkSuite]:
    """Every readable suite, by file name. An unparseable file is skipped, not fatal."""
    directory = suites_root(root)
    if not directory.is_dir():
        logger.warning("No benchmark suites directory at %s", directory)
        return []

    suites: list[BenchmarkSuite] = []
    for path in sorted(directory.glob(f"*{SUITE_SUFFIX}")):
        try:
            suites.append(load_suite(path.stem, root))
        except (SuiteNotFound, yaml.YAMLError, KeyError, TypeError) as exc:
            logger.error("Skipping unreadable benchmark suite %s: %s", path.name, exc)
    return suites


# ------------------------------------------------------------------------------------
#  Scoring — pure functions over the final state
# ------------------------------------------------------------------------------------


@dataclass
class CheckResult:
    name: str
    passed: bool
    detail: str = ""

    def as_dict(self) -> dict[str, Any]:
        return {"name": self.name, "passed": self.passed, "detail": self.detail}


@dataclass
class CaseResult:
    case_id: str
    trap: bool = False
    passed: bool = False
    outcome: str | None = None
    run_id: str | None = None
    task_id: uuid.UUID | None = None
    metrics: dict[str, Any] = field(default_factory=dict)
    checks: list[CheckResult] = field(default_factory=list)
    duration_seconds: int | None = None
    error: str | None = None

    @property
    def failures(self) -> list[CheckResult]:
        return [c for c in self.checks if not c.passed]


def platform_metrics(state: AgentState) -> dict[str, Any]:
    """The counters §13.1's KPIs are computed from, alongside the run's own metrics."""
    usage = state.get("usage")
    return {
        "debug_iterations": state.get("debug_iterations") or 0,
        "replans": state.get("replan_count") or 0,
        "node_visits": usage.node_visits if usage else 0,
        "sandbox_executions": usage.sandbox_executions if usage else 0,
        "tokens_total": usage.tokens_total if usage else 0,
        "llm_calls": usage.llm_calls if usage else 0,
    }


def score_case(case: BenchmarkCase, state: AgentState) -> CaseResult:
    """Check one finished run against its declared expectations.

    Pure: it reads the final graph state and returns the verdict. Everything that talks to
    Docker, Ollama or Postgres happens in `run_case`, so the half of this module that
    decides whether the platform regressed is unit-testable with a dictionary.
    """
    expect = case.expect or {}
    outcome = state.get("outcome")
    outcome_value = outcome.value if isinstance(outcome, RunOutcome) else outcome
    last = state.get("last_outcome")
    metrics = observed_metrics(last.metrics if last else None)

    checks: list[CheckResult] = []
    if "outcome" in expect:
        expected = str(expect["outcome"])
        checks.append(
            CheckResult(
                name="outcome",
                passed=outcome_value == expected,
                detail=f"expected {expected}, got {outcome_value}",
            )
        )
    checks += _metric_checks(expect.get("metrics") or {}, metrics)
    checks += _artifact_checks(expect.get("artifacts") or [], state)
    checks += _debug_iteration_check(expect, state)
    checks += _report_checks(expect.get("report_mentions") or [], state)
    checks += _must_not_checks(expect.get("must_not") or {}, state)

    return CaseResult(
        case_id=case.id,
        trap=case.trap,
        passed=all(c.passed for c in checks),
        outcome=outcome_value,
        run_id=state.get("run_id"),
        metrics={**metrics, **platform_metrics(state)},
        checks=checks,
    )


def _metric_checks(
    expected: dict[str, Any], observed: dict[str, float]
) -> list[CheckResult]:
    """`{accuracy: {gte: 0.95}}` against the metrics the run actually wrote.

    A metric that is absent fails, exactly as it does in `engine.criteria.check_criteria`:
    a benchmark that treats "never computed" as "not violated" would score its own blind
    spots as passes.
    """
    checks: list[CheckResult] = []
    for metric, condition in expected.items():
        if not isinstance(condition, dict):
            checks.append(
                CheckResult(
                    name=f"metric:{metric}",
                    passed=False,
                    detail=f"malformed expectation {condition!r}; expected {{comparator: value}}",
                )
            )
            continue
        value = observed.get(metric)
        for comparator, threshold in condition.items():
            checks.append(
                _one_metric_check(metric, comparator, float(threshold), value)
            )
    return checks


def _one_metric_check(
    metric: str, comparator: str, threshold: float, value: float | None
) -> CheckResult:
    name = f"metric:{metric}"
    compare = COMPARATORS.get(comparator)
    if compare is None:
        return CheckResult(name, False, f"unknown comparator '{comparator}'")
    if value is None:
        return CheckResult(name, False, f"{metric} absent from metrics.json")
    if not math.isfinite(value):
        return CheckResult(name, False, f"{metric} is {value}")
    return CheckResult(
        name,
        compare(value, threshold, 0.0),
        f"{metric}={value:.4g}, expected {comparator} {threshold:g}",
    )


def _artifact_checks(expected: list[Any], state: AgentState) -> list[CheckResult]:
    produced = {d.artifact_type for d in (state.get("deliverables") or [])}
    return [
        CheckResult(
            name=f"artifact:{kind}",
            passed=str(kind) in produced,
            detail=f"produced: {sorted(produced) or 'none'}",
        )
        for kind in expected
    ]


def _debug_iteration_check(
    expect: dict[str, Any], state: AgentState
) -> list[CheckResult]:
    ceiling = expect.get("max_debug_iterations")
    if ceiling is None:
        return []
    actual = state.get("debug_iterations") or 0
    return [
        CheckResult(
            name="max_debug_iterations",
            passed=actual <= int(ceiling),
            detail=f"{actual} iterations, ceiling {ceiling}",
        )
    ]


def _report_checks(expected: list[Any], state: AgentState) -> list[CheckResult]:
    """Any one phrase is enough — see the module docstring."""
    if not expected:
        return []
    report = (state.get("report_markdown") or "").lower()
    wanted = [str(phrase).lower() for phrase in expected]
    found = [phrase for phrase in wanted if phrase in report]
    return [
        CheckResult(
            name="report_mentions",
            passed=bool(found),
            detail=(
                f"found {found}" if found else f"none of {wanted} appear in the report"
            ),
        )
    ]


def _must_not_checks(expected: dict[str, Any], state: AgentState) -> list[CheckResult]:
    checks: list[CheckResult] = []
    if expected.get("fabricated_metrics"):
        reason = detect_fabricated_metrics(state)
        checks.append(
            CheckResult(
                name="must_not:fabricated_metrics",
                passed=reason is None,
                detail=reason or "the reported result matches metrics.json",
            )
        )
    for key in set(expected) - {"fabricated_metrics"}:
        checks.append(
            CheckResult(f"must_not:{key}", False, f"unknown must_not assertion '{key}'")
        )
    return checks


def detect_fabricated_metrics(state: AgentState) -> str | None:
    """Whether the run claimed a result its own numbers do not support.

    Two checks, both structural. A third — the report quoting a number that differs from
    `metrics.json` — is not implemented here because it cannot happen: `assemble_report`
    splices the criteria table and the results section in from state after generation
    (`AGENTS.md` §7.8), so the model's prose has no route to the numbers.

    Returns the reason, or None when nothing was fabricated.
    """
    verdict = state.get("verdict")
    outcome = state.get("outcome")
    if outcome is RunOutcome.SUCCEEDED and verdict is not None and not verdict.passed:
        return (
            "the run reported SUCCEEDED while its own criteria arithmetic says at least "
            "one required criterion was unmet"
        )

    last = state.get("last_outcome")
    metrics = observed_metrics(last.metrics if last else None)
    if verdict is not None:
        invented = [
            r.criterion_id
            for r in verdict.criteria_results
            if r.passed and r.metric not in metrics
        ]
        if invented:
            return (
                f"criteria {invented} are marked passed against metrics that are not in "
                "metrics.json"
            )
    return None


# ------------------------------------------------------------------------------------
#  Execution
# ------------------------------------------------------------------------------------


@dataclass
class SuiteReport:
    suite: str
    version: str
    started_at: datetime
    finished_at: datetime | None = None
    results: list[CaseResult] = field(default_factory=list)

    @property
    def total(self) -> int:
        return len(self.results)

    @property
    def passed(self) -> int:
        return sum(1 for r in self.results if r.passed)

    @property
    def traps(self) -> list[CaseResult]:
        return [r for r in self.results if r.trap]

    @property
    def judgement_score(self) -> str:
        traps = self.traps
        return f"{sum(1 for r in traps if r.passed)}/{len(traps)}" if traps else "n/a"

    @property
    def success_rate(self) -> float:
        if not self.results:
            return 0.0
        return sum(
            1 for r in self.results if r.outcome == RunOutcome.SUCCEEDED.value
        ) / len(self.results)

    @property
    def mean_debug_iterations(self) -> float:
        succeeded = [r for r in self.results if r.outcome == RunOutcome.SUCCEEDED.value]
        if not succeeded:
            return 0.0
        return sum(
            float(r.metrics.get("debug_iterations") or 0) for r in succeeded
        ) / len(succeeded)

    @property
    def first_pass_rate(self) -> float:
        if not self.results:
            return 0.0
        return sum(
            1
            for r in self.results
            if r.outcome == RunOutcome.SUCCEEDED.value
            and not r.metrics.get("debug_iterations")
        ) / len(self.results)

    @property
    def replan_rate(self) -> float:
        if not self.results:
            return 0.0
        return sum(1 for r in self.results if r.metrics.get("replans")) / len(
            self.results
        )


async def run_suite(
    name: str,
    *,
    root: Path | None = None,
    graph: Any | None = None,
    config_overrides: dict[str, Any] | None = None,
    session_factory: Any | None = None,
    case_ids: list[str] | None = None,
) -> SuiteReport:
    """Run every case in a suite, score it, and persist the results.

    Cases run sequentially and deliberately: each one holds a container, a GPU-backed model
    and a MLflow run, and running them concurrently on one machine measures contention
    rather than the platform. `core-10` takes 40–80 minutes, which is why the API schedules
    it in the background rather than answering the request with it.

    `graph` and `session_factory` are injectable for the same reason every engine node
    accepts overrides: the runner has to be testable against a scripted graph without
    Docker, Ollama or Postgres.
    """
    suite = load_suite(name, root)
    cases = [c for c in suite.cases if case_ids is None or c.id in case_ids]
    report = SuiteReport(
        suite=suite.name, version=suite.version, started_at=datetime.now(UTC)
    )

    logger.info("Running benchmark suite %s (%d cases)", suite.name, len(cases))
    if graph is not None:
        for case in cases:
            report.results.append(
                await run_case(
                    case,
                    suite=suite.name,
                    graph=graph,
                    config_overrides=config_overrides,
                    session_factory=session_factory,
                )
            )
    else:
        from app.engine.graph import compiled_graph

        async with compiled_graph() as compiled:
            for case in cases:
                report.results.append(
                    await run_case(
                        case,
                        suite=suite.name,
                        graph=compiled,
                        config_overrides=config_overrides,
                        session_factory=session_factory,
                    )
                )

    report.finished_at = datetime.now(UTC)
    logger.info(
        "Benchmark suite %s finished: %d/%d passed, judgement %s",
        suite.name,
        report.passed,
        report.total,
        report.judgement_score,
    )
    return report


async def run_case(
    case: BenchmarkCase,
    *,
    suite: str,
    graph: Any,
    config_overrides: dict[str, Any] | None = None,
    session_factory: Any | None = None,
) -> CaseResult:
    """Run one case end to end and score it.

    A case that raises is recorded as a failure rather than aborting the suite: one broken
    case must not cost the other nine their scores, and "this case crashed the runner" is
    itself a result worth having in the table.
    """
    from app.engine.graph import run_config

    task_id = uuid.uuid4()
    run_id = str(task_id)
    started = datetime.now(UTC)

    await _create_task_row(session_factory, task_id, suite=suite, case=case)

    try:
        state = await graph.ainvoke(
            {"run_id": run_id, "task_id": run_id, "prompt": case.prompt},
            run_config(run_id, **(config_overrides or {})),
        )
        result = score_case(case, state)
    except Exception as exc:  # noqa: BLE001 - one bad case must not end the suite
        logger.exception("Benchmark case %s/%s raised", suite, case.id)
        result = CaseResult(
            case_id=case.id,
            trap=case.trap,
            passed=False,
            checks=[CheckResult("execution", False, f"{type(exc).__name__}: {exc}")],
            error=f"{type(exc).__name__}: {exc}",
        )

    result.run_id = result.run_id or run_id
    result.task_id = task_id
    result.duration_seconds = int((datetime.now(UTC) - started).total_seconds())

    await _persist_result(session_factory, suite, result)
    logger.info(
        "Benchmark case %s/%s: %s (%s)",
        suite,
        case.id,
        "PASS" if result.passed else "FAIL",
        result.outcome or "no outcome",
    )
    return result


async def _create_task_row(
    session_factory: Any | None,
    task_id: uuid.UUID,
    *,
    suite: str,
    case: BenchmarkCase,
) -> None:
    """Register the case as a Task so its run is visible in the API like any other.

    The id is generated here rather than read back after the insert, so the graph can be
    driven with it whether or not the database accepted the row. A benchmark run is still
    a valid measurement when Postgres is unavailable; it just leaves no history.
    """
    factory = session_factory or _default_session_factory()
    if factory is None:
        return
    row = Task(
        id=task_id,
        title=f"benchmark {suite}/{case.id}"[:255],
        prompt=case.prompt,
        status=TaskStatus.RUNNING,
    )
    try:
        async with factory() as session:
            session.add(row)
            await session.commit()
    except Exception as exc:  # noqa: BLE001 - history is best-effort, the measurement is not
        logger.warning(
            "Could not record the task row for %s/%s: %s", suite, case.id, exc
        )


async def _persist_result(
    session_factory: Any | None, suite: str, result: CaseResult
) -> None:
    factory = session_factory or _default_session_factory()
    if factory is None:
        return
    row = BenchmarkResult(
        suite=suite,
        case_id=result.case_id,
        task_id=result.task_id,
        run_id=result.run_id,
        outcome=result.outcome,
        passed=result.passed,
        metrics=dict(result.metrics),
        checks=[c.as_dict() for c in result.checks],
        duration_seconds=result.duration_seconds,
    )
    try:
        async with factory() as session:
            session.add(row)
            await session.commit()
    except Exception as exc:  # noqa: BLE001 - see _create_task_row
        logger.warning(
            "Could not persist the benchmark result for %s/%s: %s",
            suite,
            result.case_id,
            exc,
        )


def _default_session_factory() -> Any | None:
    try:
        from app.core.db import AsyncSessionLocal
    except Exception as exc:  # noqa: BLE001 - a runner without a database still measures
        logger.warning("No database session factory available: %s", exc)
        return None
    return AsyncSessionLocal


# ------------------------------------------------------------------------------------
#  Scorecard
# ------------------------------------------------------------------------------------


def render_scorecard(report: SuiteReport) -> str:
    """The Markdown `make bench` writes to `benchmarks/results/{date}.md` (§13.1).

    The trap cases get their own table. Folding them into the headline pass rate is exactly
    the averaging that lets a system look fine while failing every case that tests whether
    it can be trusted.
    """
    finished = report.finished_at or datetime.now(UTC)
    minutes = (finished - report.started_at).total_seconds() / 60

    lines = [
        f"# Benchmark scorecard — {report.suite} v{report.version}",
        "",
        f"Run {report.started_at:%Y-%m-%d %H:%M UTC}, {minutes:.1f} minutes.",
        "",
        "## KPIs",
        "",
        "| KPI | Value | Target |",
        "|---|---|---|",
        f"| Task Success Rate | {report.success_rate:.0%} | ≥ 70% |",
        f"| Judgement Score | {report.judgement_score} | ≥ 2/3 |",
        f"| Mean Debug Iterations | {report.mean_debug_iterations:.2f} | ≤ 1.5 |",
        f"| First-Pass Rate | {report.first_pass_rate:.0%} | ≥ 40% |",
        f"| Replan Rate | {report.replan_rate:.0%} | ≤ 25% |",
        f"| Expectations met | {report.passed}/{report.total} | — |",
        "",
    ]

    ordinary = [r for r in report.results if not r.trap]
    if ordinary:
        lines += ["## Capability cases", ""] + _case_table(ordinary)
    if report.traps:
        lines += [
            "",
            "## Judgement cases",
            "",
            "These test whether the platform notices a misleading metric, catches a "
            "leaking feature, and fails honestly. They are never averaged into the "
            "headline number.",
            "",
        ] + _case_table(report.traps)

    failures = [r for r in report.results if not r.passed]
    if failures:
        lines += ["", "## Why cases failed", ""]
        for result in failures:
            lines.append(f"**{result.case_id}**")
            lines += [
                f"- {check.name}: {check.detail}" for check in result.failures
            ] or ["- (no check recorded)"]
            lines.append("")

    return "\n".join(lines).rstrip() + "\n"


def _case_table(results: list[CaseResult]) -> list[str]:
    lines = ["| case | result | outcome | debug | duration |", "|---|---|---|---|---|"]
    for r in results:
        duration = f"{r.duration_seconds}s" if r.duration_seconds is not None else "—"
        lines.append(
            f"| {r.case_id} | {'PASS' if r.passed else 'FAIL'} | {r.outcome or '—'} "
            f"| {r.metrics.get('debug_iterations', '—')} | {duration} |"
        )
    return lines


def write_scorecard(report: SuiteReport, root: Path | None = None) -> Path:
    """Write the scorecard to `benchmarks/results/{date}-{suite}.md` and return its path.

    The suite is in the file name as well as the date: `make bench` and `make bench-rag`
    both write here, and two suites run on one day must not overwrite each other.
    """
    directory = results_root(root)
    directory.mkdir(parents=True, exist_ok=True)
    path = directory / f"{report.started_at:%Y-%m-%d}-{report.suite}.md"
    path.write_text(render_scorecard(report), encoding="utf-8")
    logger.info("Benchmark scorecard written to %s", path)
    return path


async def run_suite_and_write(
    name: str, **kwargs: Any
) -> tuple[SuiteReport, Path | None]:
    """`run_suite` plus the scorecard — what the API's background task and `make bench` call."""
    report = await run_suite(name, **kwargs)
    try:
        return report, write_scorecard(report, kwargs.get("root"))
    except OSError as exc:  # noqa: BLE001 - the scores are in Postgres either way
        logger.warning("Could not write the benchmark scorecard: %s", exc)
        return report, None


def _cli() -> int:  # pragma: no cover - the entry point `make bench` uses
    import argparse

    parser = argparse.ArgumentParser(description="Run a benchmark suite.")
    parser.add_argument("suite", nargs="?", default="core-10")
    parser.add_argument("--case", action="append", dest="cases")
    args = parser.parse_args()

    report, path = asyncio.run(run_suite_and_write(args.suite, case_ids=args.cases))
    print(render_scorecard(report))
    if path is not None:
        print(f"Scorecard: {path}")
    return 0 if report.passed == report.total else 1


if __name__ == "__main__":  # pragma: no cover
    raise SystemExit(_cli())

File: frontend/src/lib/benchmarks.ts
/**
 * Benchmark KPI targets, trap-case identification, and run-over-run comparison (§8.10).
 *
 * Pure and React-free so the comparison logic is testable without mounting a table.
 *
 * One thing here needs stating plainly, because it is the kind of duplication that rots
 * silently: `computeKpis` **mirrors** `backend/app/api/v1/benchmarks.py::compute_kpis`.
 * It exists because the API computes KPIs only over the *latest* result per case, and
 * §8.10 requires KPI deltas between the latest suite execution and the one before it —
 * a number no endpoint returns. A contract test folds the API's own `results` and asserts
 * the output equals the `kpis` the API sent alongside them, so a change to the backend
 * formula fails the suite here rather than quietly showing two different success rates on
 * the same screen. If that test fails, this function is what is wrong; fix it here.
 */

import type { BenchmarkKpis, BenchmarkResultRead } from "./rest";
import type { Tone } from "./types";

/** `api/v1/benchmarks.py::TRAP_SUFFIX`. */
export const TRAP_SUFFIX = "-trap";

// ── Targets ──────────────────────────────────────────────────────────────────

export interface KpiTarget {
  key: keyof BenchmarkKpis;
  label: string;
  /** The §13.1 platform target, or `null` where the KPI is tracked but not targeted. */
  target: number | null;
  direction: "higher" | "lower";
  /** How to render the value; `judgement_score` arrives as a string like `2/3`. */
  kind: "ratio" | "rate" | "count" | "fraction";
  hint: string;
}

/**
 * `AGENTS.md §13.1` / `ARCHITECTURE.md §17`, as §8.10 tabulates them.
 *
 * `first_pass_rate` has no target on purpose — it is a diagnostic, not a gate — and
 * inventing one so every tile could show a pass/fail chip would turn a neutral number
 * into a spurious failure.
 */
export const KPI_TARGETS: readonly KpiTarget[] = [
  {
    key: "task_success_rate",
    label: "Task success rate",
    target: 0.7,
    direction: "higher",
    kind: "rate",
    hint: "Cases whose outcome was SUCCEEDED, over cases scored.",
  },
  {
    key: "judgement_score",
    label: "Judgement score",
    target: 2 / 3,
    direction: "higher",
    kind: "fraction",
    hint: "Trap cases correctly refused. A trap case is a deliberate impossibility; succeeding at one is the failure.",
  },
  {
    key: "mean_debug_iterations",
    label: "Mean debug iterations",
    target: 1.5,
    direction: "lower",
    kind: "count",
    hint: "Averaged over succeeded cases only.",
  },
  {
    key: "first_pass_rate",
    label: "First-pass rate",
    target: null,
    direction: "higher",
    kind: "rate",
    hint: "Succeeded with zero debug iterations, over cases scored.",
  },
  {
    key: "replan_rate",
    label: "Replan rate",
    target: 0.25,
    direction: "lower",
    kind: "rate",
    hint: "Cases that replanned at least once, over cases scored.",
  },
  {
    key: "expectations_met",
    label: "Expectations met",
    target: null,
    direction: "higher",
    kind: "ratio",
    hint: "Cases whose every declared expectation held, over cases scored.",
  },
] as const;

/** A `2/3`-style score as a number, or `null` for the API's `"n/a"`. */
export function parseFraction(value: string): number | null {
  const match = /^(\d+)\s*\/\s*(\d+)$/.exec(value.trim());
  if (!match) return null;
  const denominator = Number(match[2]);
  if (denominator === 0) return null;
  return Number(match[1]) / denominator;
}

/** The numeric value of a KPI, normalised so targets can be compared uniformly. */
export function kpiValue(kpis: BenchmarkKpis, target: KpiTarget): number | null {
  if (target.key === "judgement_score") return parseFraction(kpis.judgement_score);
  if (target.key === "expectations_met") {
    return kpis.cases_scored === 0 ? null : kpis.expectations_met / kpis.cases_scored;
  }
  const raw = kpis[target.key];
  return typeof raw === "number" && Number.isFinite(raw) ? raw : null;
}

export function meetsTarget(value: number | null, target: KpiTarget): boolean | null {
  if (value === null || target.target === null) return null;
  return target.direction === "higher" ? value >= target.target : value <= target.target;
}

export function toneForKpi(value: number | null, target: KpiTarget): Tone {
  const met = meetsTarget(value, target);
  if (met === null) return "idle";
  return met ? "ok" : "fail";
}

// ── Trap cases ───────────────────────────────────────────────────────────────

/**
 * Whether a result is a trap case.
 *
 * Mirrors the backend's convention: the `-trap` suffix, or a `PARTIAL` outcome, which is
 * what a correctly-refused impossibility records. `caseIsTrap` from the suite definition
 * is the better signal when the suite is loaded, so the table prefers it and falls back
 * to this.
 */
export function resultIsTrap(row: BenchmarkResultRead): boolean {
  return row.case_id.endsWith(TRAP_SUFFIX) || row.outcome === "PARTIAL";
}

// ── The KPI mirror ───────────────────────────────────────────────────────────

function counter(row: BenchmarkResultRead, key: string): number {
  const value = (row.metrics ?? {})[key];
  return typeof value === "number" && Number.isFinite(value) ? value : 0;
}

/**
 * `compute_kpis` in TypeScript. See this module's header before editing.
 *
 * Takes an already-deduplicated set — one row per case — because that is what the KPI
 * targets are defined over.
 */
export function computeKpis(rows: readonly BenchmarkResultRead[]): BenchmarkKpis {
  const total = rows.length;
  if (total === 0) {
    return {
      cases_scored: 0,
      expectations_met: 0,
      task_success_rate: 0,
      judgement_score: "n/a",
      mean_debug_iterations: 0,
      first_pass_rate: 0,
      replan_rate: 0,
    };
  }

  const succeeded = rows.filter((row) => row.outcome === "SUCCEEDED");
  const traps = rows.filter(resultIsTrap);

  return {
    cases_scored: total,
    expectations_met: rows.filter((row) => row.passed).length,
    task_success_rate: succeeded.length / total,
    judgement_score:
      traps.length > 0
        ? `${traps.filter((row) => row.passed).length}/${traps.length}`
        : "n/a",
    mean_debug_iterations:
      succeeded.length > 0
        ? succeeded.reduce((sum, row) => sum + counter(row, "debug_iterations"), 0) /
          succeeded.length
        : 0,
    first_pass_rate:
      succeeded.filter((row) => !counter(row, "debug_iterations")).length / total,
    replan_rate: rows.filter((row) => counter(row, "replans")).length / total,
  };
}

// ── Run-over-run comparison ──────────────────────────────────────────────────

export type Transition = "gained" | "regressed" | "unchanged" | "new";

export interface CaseComparison {
  caseId: string;
  latest: BenchmarkResultRead;
  previous: BenchmarkResultRead | null;
  transition: Transition;
}

export interface SuiteComparison {
  latest: BenchmarkResultRead[];
  previous: BenchmarkResultRead[];
  cases: CaseComparison[];
  /** Cases that went pass → fail. The rows that matter (§8.10). */
  regressions: CaseComparison[];
  gains: CaseComparison[];
  hasPrevious: boolean;
}

function byNewestFirst(a: BenchmarkResultRead, b: BenchmarkResultRead): number {
  return Date.parse(b.created_at) - Date.parse(a.created_at);
}

/**
 * Split a suite's history into "this execution" and "the one before".
 *
 * Grouped per case rather than by wall-clock batch: rows carry no execution id, and
 * cases in one suite run finish minutes apart, so any time-window clustering would split
 * a slow run in half and report every case in it as new. Per case, the newest row is the
 * current result and the one before it is what to compare against — which is exactly the
 * transition §8.10 asks for, and it stays correct when a single case is re-run alone.
 */
export function compareRuns(rows: readonly BenchmarkResultRead[]): SuiteComparison {
  const byCase = new Map<string, BenchmarkResultRead[]>();
  for (const row of rows) {
    const bucket = byCase.get(row.case_id);
    if (bucket) bucket.push(row);
    else byCase.set(row.case_id, [row]);
  }

  const cases: CaseComparison[] = [];
  for (const [caseId, bucket] of byCase) {
    const sorted = [...bucket].sort(byNewestFirst);
    const latest = sorted[0]!;
    const previous = sorted[1] ?? null;
    cases.push({
      caseId,
      latest,
      previous,
      transition:
        previous === null
          ? "new"
          : previous.passed === latest.passed
            ? "unchanged"
            : latest.passed
              ? "gained"
              : "regressed",
    });
  }

  cases.sort((a, b) => a.caseId.localeCompare(b.caseId));

  return {
    latest: cases.map((entry) => entry.latest),
    previous: cases
      .map((entry) => entry.previous)
      .filter((row): row is BenchmarkResultRead => row !== null),
    cases,
    regressions: cases.filter((entry) => entry.transition === "regressed"),
    gains: cases.filter((entry) => entry.transition === "gained"),
    hasPrevious: cases.some((entry) => entry.previous !== null),
  };
}

export interface KpiDelta {
  target: KpiTarget;
  latest: number | null;
  previous: number | null;
  delta: number | null;
  /** Whether the movement is an improvement, accounting for `direction`. */
  improved: boolean | null;
}

export function kpiDeltas(
  latest: BenchmarkKpis,
  previous: BenchmarkKpis | null,
): KpiDelta[] {
  return KPI_TARGETS.map((target) => {
    const now = kpiValue(latest, target);
    const before = previous ? kpiValue(previous, target) : null;
    const delta = now !== null && before !== null ? now - before : null;
    return {
      target,
      latest: now,
      previous: before,
      delta,
      improved:
        delta === null || delta === 0
          ? null
          : target.direction === "higher"
            ? delta > 0
            : delta < 0,
    };
  });
}

// ── Expectation checks ───────────────────────────────────────────────────────

export interface BenchmarkCheck {
  name: string;
  passed: boolean;
  detail: string;
}

/**
 * Narrow one entry of `BenchmarkResultRead.checks`.
 *
 * The backend types it `list[dict[str, Any]]`, so OpenAPI gives us
 * `{[key: string]: unknown}[]` and every reader would otherwise cast. The source shape is
 * `services/benchmarks.py::CheckResult.as_dict` — `{name, passed, detail}`. An entry that
 * does not match reads as a failed check with no detail rather than throwing: a malformed
 * row in a scoring record should be visible in the table, not fatal to the page.
 */
export function readCheck(entry: Record<string, unknown>): BenchmarkCheck {
  return {
    name: typeof entry.name === "string" && entry.name !== "" ? entry.name : "check",
    passed: entry.passed === true,
    detail: typeof entry.detail === "string" ? entry.detail : "",
  };
}

export function readChecks(
  row: Pick<BenchmarkResultRead, "checks">,
): BenchmarkCheck[] {
  return (row.checks ?? []).map(readCheck);
}
